### PR TITLE
Feature/context decorator

### PR DESCRIPTION
**Pull request by JanTrotnow**

### CODE REVIEW

File: spec/support/mocks/states/context-c.ts [hunk 1]
```diff
@@ -0,0 +1,31 @@
+import { inject, injectable, optional } from "inversify";
+import { TranslateHelper } from "../../../../src/components/i18n/public-interfaces";
+import { Logger } from "../../../../src/components/root/public-interfaces";
+import { BaseState } from "../../../../src/components/state-machine/base-state";
+import { clearContext } from "../../../../src/components/state-machine/clear-context-decorator";
+import { State, Transitionable } from "../../../../src/components/state-machine/public-interfaces";
+import { stayInContext } from "../../../../src/components/state-machine/stay-in-context-decorator";
+import { ResponseFactory } from "../../../../src/components/unifier/public-interfaces";
+
+@stayInContext((current, next, context, history) => true)
```
Comment (antoniusostermann) [inline spec/support/mocks/states/context-c.ts]: Is it possible to emit the callback of both functions? The callback you provided should be the default case.

File: spec/components/state-machine/context.spec.ts [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+        await doStateTransitions(this.stateMachine, ["ContextAState", "ContextCState"]);
+      });
+
+      it("removes ContextAState from context", async function(this: CurrentThisContext) {
+        await this.stateMachine.handleIntent("exampleCIntent");
+        const contextStates = await this.getContextStates();
+        expect(contextStates.length).toBe(0);
+      });
+    });
+
```
Comment (antoniusostermann) [inline spec/components/state-machine/context.spec.ts]: Please add a spec to test if the callback functions in `@stayInContext` and `@clearContext` get the correct arguments.

File: src/components/state-machine/clear-context-decorator.ts [hunk 3]
```diff
@@ -0,0 +1,11 @@
+import { State } from "../state-machine/public-interfaces";
```
Comment (antoniusostermann) [inline src/components/state-machine/clear-context-decorator.ts]: Can we put all state machine decoratos in a seperate subfolder `state-machine/decorators`? This also includes your `filter` contribution.

File: src/components/state-machine/state-machine.ts [hunk 4]
```diff
@@ -4,28 +4,36 @@ import { Logger } from "../root/public-interfaces";
...
+  public async handleIntent(requestedIntent: intent, ...args: any[]) {
     const currentState = await this.getCurrentState();
-
-    const intentMethod = this.deriveIntentMethod(intent);
+    const intentMethod = this.deriveIntentMethod(requestedIntent);
     this.intentHistory.push({ stateName: currentState.name, intentMethodName: intentMethod });
     this.logger.info("Handling intent '" + intentMethod + "' on state " + currentState.name);
 
+    // execute clearContext callback if decorator is present
+    const clearContextCallbackFn = this.retrieveClearContextCallbackFromMetadata((currentState.instance as any).constructor);
```
Comment (antoniusostermann) [inline src/components/state-machine/state-machine.ts]: Do you really need this `any`-typecast?

File: src/components/state-machine/state-machine.ts [hunk 5]
```diff
@@ -4,28 +4,36 @@ import { Logger } from "../root/public-interfaces";
...
+    if (clearContextCallbackFn && clearContextCallbackFn(currentState, (await this.getContextStates()).map(cState => cState.name), this.intentHistory)) {
```
Comment (antoniusostermann) [inline src/components/state-machine/state-machine.ts]: Above, you also have the `await this.getCurrentState()` call. Since this and the `getContextState` call can run in parallel without side effects, it would be pretty nice to sum them up above (`const [currentState, contextStates] = await Promise.all(...)`).

File: src/components/state-machine/state-machine.ts [hunk 6]
```diff
@@ -69,8 +79,16 @@ export class StateMachine implements Transitionable {
           .withArguments(currentState.instance, currentState.name, intentMethod, this, ...args)
           .runWithResultset();
       } else {
-        // -> Intent does not exist on state class, so call unhandledGenericIntent instead
-        await this.handleIntent(GenericIntent.Unhandled, intentMethod, ...args);
+        const contextStates = await this.getContextStates();
```
Comment (antoniusostermann) [inline src/components/state-machine/state-machine.ts]: Here, you could reuse the resolved promise (see comment above)

File: src/components/state-machine/state-machine.ts [hunk 7]
```diff
@@ -80,13 +98,35 @@ export class StateMachine implements Transitionable {
 
   public async transitionTo(state: string) {
     if (this.stateNames.indexOf(state) === -1) throw Error("Cannot transition to " + state + ": State does not exist!");
+    let contextStates = await this.getContextStates();
+
+    // add current state to context if context meta data is present
+    const currentState = await this.getCurrentState();
```
Comment (antoniusostermann) [inline src/components/state-machine/state-machine.ts]: Can run in parallel with instruction above

File: src/components/state-machine/state-machine.ts [hunk 8]
```diff
@@ -69,8 +79,16 @@ export class StateMachine implements Transitionable {
           .withArguments(currentState.instance, currentState.name, intentMethod, this, ...args)
           .runWithResultset();
       } else {
-        // -> Intent does not exist on state class, so call unhandledGenericIntent instead
-        await this.handleIntent(GenericIntent.Unhandled, intentMethod, ...args);
+        const contextStates = await this.getContextStates();
+        const fittingState = contextStates.find(state => typeof state.instance[intentMethod] === "function");
+
+        if (fittingState) {
+          await this.transitionTo((fittingState as { instance: State.Required; name: string }).name);
```
Comment (antoniusostermann) [inline src/components/state-machine/state-machine.ts]: Why do you need this type cast?

File: src/components/state-machine/state-machine.ts [hunk 9]
```diff
@@ -4,28 +4,36 @@ import { Logger } from "../root/public-interfaces";
...
 import { State, Transitionable } from "./public-interfaces";
+import { stayInContextMetadataKey } from "./stay-in-context-decorator";
 
 @injectable()
 export class StateMachine implements Transitionable {
   public intentHistory: Array<{ stateName: string; intentMethodName: string }> = [];
 
   constructor(
     @inject("core:state-machine:current-state-provider") private getCurrentState: () => Promise<{ instance: State.Required; name: string }>,
+    @inject("core:state-machine:context-states-provider") private getContextStates: () => Promise<Array<{ instance: State.Required; name: string }>>,
```
Comment (antoniusostermann) [inline src/components/state-machine/state-machine.ts]: Please create an interface for your provider. You could to the same for `getCurrentState` to clean up the old code, too.

File: src/components/state-machine/state-machine.ts [hunk 10]
```diff
@@ -69,8 +79,16 @@ export class StateMachine implements Transitionable {
...
           .runWithResultset();
       } else {
-        // -> Intent does not exist on state class, so call unhandledGenericIntent instead
-        await this.handleIntent(GenericIntent.Unhandled, intentMethod, ...args);
+        const contextStates = await this.getContextStates();
+        const fittingState = contextStates.find(state => typeof state.instance[intentMethod] === "function");
+
+        if (fittingState) {
+          await this.transitionTo((fittingState as { instance: State.Required; name: string }).name);
+          await this.handleIntent(intentMethod);
```
Comment (antoniusostermann) [inline src/components/state-machine/state-machine.ts]: Don't you need to pass `..args`?

File: spec/components/state-machine/context.spec.ts (hunk 2 above)
Comment (antoniusostermann) [reply spec/components/state-machine/context.spec.ts]: Please also add a test if additional arguments are passed through contextual derived intent methods, see below.

File: src/components/state-machine/state-machine.ts [hunk 11]
```diff
@@ -80,13 +98,35 @@ export class StateMachine implements Transitionable {
...
   public async transitionTo(state: string) {
     if (this.stateNames.indexOf(state) === -1) throw Error("Cannot transition to " + state + ": State does not exist!");
+    let contextStates = await this.getContextStates();
+
+    // add current state to context if context meta data is present
+    const currentState = await this.getCurrentState();
+    const stayInContextCallbackFn = this.retrieveStayInContextCallbackFromMetadata((currentState.instance as any).constructor);
+
+    // add to context if not present already
+    if (stayInContextCallbackFn && contextStates.map(cState => cState.name).indexOf(currentState.name) === -1) {
```
Comment (antoniusostermann) [inline src/components/state-machine/state-machine.ts]: This won't "bump" states in the context, will it? Consider this context: `[A, B, C, D]` - if we transition from C and have a `@stayInContext`, shouldn't the new context be like `[A, B, D, C]` (or `[C, A, B, D]`, depending on your order)? We'd also need a test for this, since none of your tests marked this issue.

File: src/components/state-machine/state-machine.ts [hunk 12]
```diff
@@ -80,13 +98,35 @@ export class StateMachine implements Transitionable {
...
+      contextStates.push(currentState);
+    }
+
+    // execute callbacks of context states and filter by result
+    contextStates = contextStates.filter(contextState =>
+      (this.retrieveStayInContextCallbackFromMetadata((contextState.instance as any).constructor) as ((...args: any[]) => boolean))(
```
Comment (antoniusostermann) [inline src/components/state-machine/state-machine.ts]: We shouldn't need these type casts if we use good method sigantures and type checks

File: src/components/state-machine/state-machine.ts [hunk 13]
```diff
@@ -119,4 +161,14 @@ export class StateMachine implements Transitionable {
   private getAfterIntentCallbacks() {
     return this.pipeFactory(componentInterfaces.afterIntent);
   }
+
+  private retrieveStayInContextCallbackFromMetadata(currentStateClass: State.Constructor): ((...args: any[]) => boolean) | undefined {
+    const metadata = Reflect.getMetadata(stayInContextMetadataKey, currentStateClass);
+    return metadata ? metadata.stayInContext : undefined;
```
Comment (antoniusostermann) [inline src/components/state-machine/state-machine.ts]: Here, you could return a callback which always resolve to true instead of `undefined` to have a "default" callback. Same applies for `clearContext`.

File: spec/components/state-machine/context.spec.ts [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+    describe("with transitioning from state A to B", function() {
+      beforeEach(async function(this: CurrentThisContext) {
+        await doStateTransitions(this.stateMachine, ["ContextAState", "ContextBState"]);
+      });
+
+      it("uses exampleAIntent of ContextAState", async function(this: CurrentThisContext) {
+        await this.stateMachine.handleIntent("exampleAIntent");
+        expect(this.stateMachine.handleIntent).toHaveBeenCalledTimes(2);
+        expect(this.stateMachine.handleIntent).not.toHaveBeenCalledWith(GenericIntent.Unhandled, "exampleAIntent");
```
[the remark this answers is not in the document]
Comment (JanTrotnow) [reply spec/components/state-machine/context.spec.ts]: ```exampleAIntent```only exists in ```ContextAState``` so there is no need to make sure it's the correct state.

File: spec/components/state-machine/context.spec.ts [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+      });
+
+      it("does not add states which stayInContext decorator callback returns false to context", async function(this: CurrentThisContext) {
+        const contextStates = await this.getContextStates();
+        expect(contextStates.map(state => state.name)).not.toContain("ContextBState");
+      });
+
+      it("uses undhandledGenericIntent of ContextAState", async function(this: CurrentThisContext) {
+        await this.stateMachine.handleIntent("exampleBIntent");
+        expect(this.stateMachine.handleIntent).toHaveBeenCalledWith(GenericIntent.Unhandled, "exampleBIntent");
```
[the remark this answers is not in the document]
Comment (JanTrotnow) [reply spec/components/state-machine/context.spec.ts]: Assuming that ```doStateTransitions```-function works properly, we're in ```ContextStateA```when ```handleIntent```is called.

File: spec/support/mocks/states/context-c.ts (hunk 1 above)
Comment (JanTrotnow) [reply spec/support/mocks/states/context-c.ts]: I added a default callback which always evaluates to true so that it's possible to simply add ```@stayInContext()```or ```@clearContext()```to a state. If you had something different in mind, feel free to leave a comment.

File: src/components/state-machine/state-machine.ts (hunk 13 above)
Comment (JanTrotnow) [reply src/components/state-machine/state-machine.ts]: I took a different approach by adding a default callback to the respective decorators because the existence of defined metadata is the criteria which is used to decide wether or not a state is added to context.
